File: old/2021/rust/01.rs
fn solve(inp: Vec<i32>) -> (u32, u32) {
    let mut part_one: u32 = 0;
    let mut part_two: u32 = 0;

    for i in 0..inp.len() - 3 {
        if inp[i+1] > inp[i] {
            part_one += 1;
        }

        let partial = inp[i+1] + inp[i+2];
        if inp[i] + partial < inp[i+3] + partial {
            part_two += 1;
        }
    }

    if inp[inp.len() - 3] > inp[inp.len() - 2] {
        part_one += 1;
    }
    if inp[inp.len() - 2] > inp[inp.len() - 1] {
        part_one += 1;
    }

    (part_one, part_two)
}

fn main() {
    let inp = std::fs::read_to_string("../input/01").unwrap().trim().to_string();
    let inp = inp.lines().map(|n| n.parse::<i32>().unwrap()).collect::<Vec<_>>();

    let (part_one, part_two) = solve(inp);
   
    println!("Part one: {}", part_one);
    println!("Part two: {}", part_two);
}

File: 2015/d01-not_quite_lisp/langs/01.hs
module Main where

part_one :: [Char] -> Int -> Int
part_one [] loc = loc
part_one (x:xs) loc | x == '(' = part_one xs (loc + 1)
                    | x == ')' = part_one xs (loc - 1)
                    | x /= '(' && x /= ')' = part_one xs loc

part_two :: [Char] -> Int -> Int -> Int
part_two [] idx loc = idx
part_two (x:xs) idx loc | loc < 0 = idx
                        | x == '(' = part_two xs (idx + 1) (loc + 1)
                        | x == ')' = part_two xs (idx + 1) (loc - 1)
                        | x /= '(' && x /= ')' = part_two xs idx loc

main :: IO ()
main = do 
    inp <- readFile "../../../input/2015/01"
    putStrLn ("Part one: " ++ (show (part_one inp 0)))
    putStrLn ("Part two: " ++ (show (part_two inp 0 0)))
